source/constructs/api/discovery_job/service.py

```diff
@@ -845,27 +979,46 @@ def __get_run_error_log(run_database: models.DiscoveryJobRunDatabase) -> str:
                               aws_session_token=credentials['SessionToken'],
                               region_name=region,
                               )
+    max_results = 1 if error_log else 5
     try:
         response = client_sfn.get_execution_history(
             executionArn=f'arn:{partition}:states:{region}:{account_id}:execution:{const.SOLUTION_NAME}-DiscoveryJob:{const.SOLUTION_NAME}-{run_database.run_id}-{run_database.id}-{run_database.uuid}',
             reverseOrder=True,
-            maxResults=1,
+            maxResults=max_results,
         )
     except client_sfn.exceptions.ExecutionDoesNotExist as e:
-        return RunDatabaseState.NOT_EXIST.value
-    if response["events"][0]["type"] == "ExecutionFailed":
-        return response["events"][0]["executionFailedEventDetails"]["cause"]
-    return ""
+        return RunDatabaseState.NOT_EXIST.value, ""
+    if error_log:
+        if response["events"][0]["type"] == "ExecutionFailed":
+            return RunDatabaseState.FAILED.value, response["events"][0]["executionFailedEventDetails"]["cause"]
+        return RunDatabaseState.FAILED.value, ""
+    else:
+        parameters = json.loads(response["events"][4]["taskScheduledEventDetails"]["parameters"])
+        result = parameters.get("MessageBody", {}).get("Result")
+        if result:
+            return result.get("State"), result.get("Message", "")
+        return RunDatabaseState.SUCCEEDED.value, ""
 
 
 def __get_cell_value(cell: dict):
-    if "VarCharValue" in cell:
-        return cell["VarCharValue"]
-    else:
-        return ""
+    return cell.get("VarCharValue", "")
 
 
-def get_report_url(run_id: int):
+def generate_report(job_id: int, run_id: int, s3_client=None, key_name=None):
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This function contains 47 lines of code, not including blank lines or lines with only comments, Python punctuation characters, identifiers, or literals. By comparison, 98% of the functions in the CodeGuru reference dataset contain fewer lines of code. Large functions might be difficult to read and have logic that is hard to understand and test.

We recommend that you simplify this function or break it into multiple functions. For example, consider extracting the code blocks on lines 1026-1056, lines 1058-1064 into separate functions.